src/components/panel/panel.js

```diff
@@ -832,6 +850,47 @@ MdPanelRef.prototype._addStyles = function() {
 };
 
 
+/**
+ * Add ARIA-specific attributes appropriate for panels.
+ */
+MdPanelRef.prototype._configureAria = function() {
+  this._panelEl.attr('tabIndex', '-1');
+
+  this._configureTrapFocus();
+};
+
+
+/**
+ * Setup the focus traps.
+ * @private
+ */
+MdPanelRef.prototype._configureTrapFocus = function() {
+  if (this._config.trapFocus) {
+    var element = this._panelEl;
+    // Set up elements before and after the panel to capture focus and
+    // redirect back into the panel.
+    this._topFocusTrap = document.createElement('div');
+    this._topFocusTrap.classList.add('_md-panel-focus-trap');
+    this._topFocusTrap.tabIndex = 0;
+
+    this._bottomFocusTrap = this._topFocusTrap.cloneNode(false);
+
+    // When focus is about to move out of the panel, we want to intercept it and redirect it
+    // back to the panel element.
+    var focusHandler = function () {
+      element.focus();
+    };
+    this._topFocusTrap.addEventListener('focus', focusHandler);
+    this._bottomFocusTrap.addEventListener('focus', focusHandler);
+
+    // The top focus trap inserted immeidately before the md-panel element (as a sibling).
+    // The bottom focus trap is inserted at the very end of the md-panel element (as a child).
+    element[0].parentNode.insertBefore(this._topFocusTrap, element[0]);
+    element.after(this._bottomFocusTrap);
```

Comment:
Didn't @ErinCoughlan fix a bug in $mdDialog where having the bottom focus trap as a child caused some bug?